### PR TITLE
Subsidy API implementation and Integration test

### PR DESCRIPTION
**Pull request by DedrickEnc**
In this PR, I implemented the subsidy API and tested all routes
The subsidy API has methods :
POST
GET
PUT
DELETE

The codes.js was updated to add more error cases, the routes.js file was updated too.

### CODE REVIEW

**Comment by DedrickEnc** [comment]
@jniles can you have a look?

File: server/config/routes.js [hunk 1]
```diff
@@ -135,7 +137,14 @@ exports.configure = function (app) {
   app.post('/references', reference.create);
   app.put('/references/:id', reference.update);
   app.delete('/references/:id', reference.remove);
-  
+
+ //API for subsidy routes crud
+  app.get('/subsidies', subsidy.list);
+  app.get('/subsidies/:id', subsidy.getSubsidy);
```
Comment (jniles) [inline server/config/routes.js]: As discussed, this should be `sudsidy.detail`.

File: server/controllers/finance/subsidy.js [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+function isEmptyObject(object) {
+  return Object.keys(object).length === 0;
+}
+
+function checkData (obj, codes){
+
+  if(isEmptyObject(obj)) { throw codes.ERR_EMPTY_BODY;}
+  if(!obj.value) { throw codes.ERR_PARAMETERS_REQUIRED;}
+  if(obj.value <= 0) { throw codes.ERR_BAD_VALUE;}
```
Comment (jniles) [inline server/controllers/finance/subsidy.js:107]: It would help to have a more specific error message here. Why is this a bad value?

File: server/controllers/finance/subsidy.js [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+function isEmptyObject(object) {
+  return Object.keys(object).length === 0;
+}
+
+function checkData (obj, codes){
+
+  if(isEmptyObject(obj)) { throw codes.ERR_EMPTY_BODY;}
+  if(!obj.value) { throw codes.ERR_PARAMETERS_REQUIRED;}
```
Comment (jniles) [inline server/controllers/finance/subsidy.js:106]: Hmm..  The ERR_PARAMETERS_REQUIRED was originally designed for HTTP parameters - query strings.  I see now that the error message is fairly generic.  It might be better to split these errors into two: 1) missing query parameters (used in @mbayopanda's patient search API) and 2) missing req.body properties for UPDATEs/CREATEs.

File: server/test/api/subsidy.js [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+var chai = require('chai');
+var chaiHttp = require('chai-http');
+var expect = chai.expect;
+chai.use(chaiHttp);
+
+var helpers = require('./helpers');
+helpers.configure(chai);
+
+var url = 'https://localhost:8080';
+var user = { username : 'superuser', password : 'superuser', project: 1};
```
Comment (jniles) [inline server/test/api/subsidy.js]: These two variables are unused.

File: server/test/api/subsidy.js [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+  var wrongSubsidy = {
+    account_id : 3626,
+    label : 'tested wrong subsidy',
+    description : 'This is for the wrong value test',
+    value : -10  
+  };
+
+  beforeEach(helpers.login(agent));
+
+  it('METHOD : POST, PATH : /subsidies, It adds a subsidy', function () {
```
Comment (jniles) [inline server/test/api/subsidy.js:30]: Good test!

File: server/test/api/subsidy.js [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        expect(res).to.have.status(201);
+        expect(res).to.be.json;
+        expect(res.body).to.not.be.empty;
+        expect(res.body.id).to.be.defined;
+        newSubsidy.id = res.body.id;
+        return agent.get('/subsidies/' + newSubsidy.id);  
+      })
+      .then(function (res){ 
+        expect(res).to.have.status(200);
+        expect(res.body).to.have.all.keys('id', 'account_id', 'label', 'description', 'value', 'created_at', 'updated_at');
```
Comment (jniles) [inline server/test/api/subsidy.js:43]: Since you reuse these keys over and over, you might want to move them into their own variable at the top of your `describe()` block.

File: server/test/api/subsidy.js [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+      })
+     .catch(helpers.handler);
+  });
+
+  it('METHOD : POST, PATH : /subsidies, It refuses to add a wrong subsidy', function () {
+    return agent.post('/subsidies')
+      .send(wrongSubsidy)
+      .then(function (res) {
```
Comment (jniles) [inline server/test/api/subsidy.js:51]: Another thing to test for: what should the error code (sent in the JSON object) be?

File: server/test/api/subsidy.js [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+        .then(function (res) {
+          expect(res).to.have.status(200);
+          expect(res).to.be.json;
+          expect(res.body).to.not.be.empty;
+          expect(res.body).to.have.length(1);
+        })
+        .catch(helpers.handler);
+    });
+  
+  it('METHOD : GET, PATH : /subsidies/:id, It returns one subsidy', function () {
```
Comment (jniles) [inline server/test/api/subsidy.js:71]: This test is a duplicate of the second half of 'It adds a new subsidy' (Line 33).  You don't need it.

File: server/test/api/subsidy.js [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+  it('METHOD : PUT, PATH : /subsidies/:id, It updates the newly added subsidy', function () {
+    var updateInfo = {value : 50};
+
+    return agent.put('/subsidies/'+ newSubsidy.id)
+      .send(updateInfo)
+      .then(function (res) {
+        expect(res).to.have.status(200);
+        expect(res).to.be.json;
+        expect(res.body.id).to.equal(newSubsidy.id);
+        expect(res.body.value).to.equal(updateInfo.value);
```
Comment (jniles) [inline server/test/api/subsidy.js:93]: It would be useful to check that other properties didn't change.

File: server/test/api/subsidy.js [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+  });
+
+   it('METHOD : DELETE, PATH : /subsidies/:id, It deletes a subsidy', function () {
+    return agent.delete('/subsidies/' + newSubsidy.id)
+      .then(function (res) {
+        expect(res).to.have.status(204);
+        // re-query the database
+        return agent.get('/subsidies/' + newSubsidy.id);
+      })
+      .then(function (res) {
```
Comment (jniles) [inline server/test/api/subsidy.js:105]: If you want to be _super_ sure that this worked correctly, you can also query `/subsidies` and check that the length is back to being the original length expected.

File: server/test/api/subsidy.js [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+ 
+  var wrongSubsidy = {
+    account_id : 3626,
+    label : 'tested wrong subsidy',
+    description : 'This is for the wrong value test',
+    value : -10  
+  };
+
+  beforeEach(helpers.login(agent));
+
```
Comment (jniles) [inline server/test/api/subsidy.js:29]: One more test to add: GET /subsidies when there are no subsidies in the database should return a 200 status and empty array.

File: server/test/api/subsidy.js [hunk 12]
```diff
@@ -0,0 +1,112 @@
+/*global describe, it, beforeEach, process*/
```
Comment (jniles) [inline server/test/api/subsidy.js]: You don't need "process" in this comment.

**Comment by jniles** [comment]
@DedrickEnc this is a very good PR.  I've made a few comments, also a few of them are semantic instead of logical.  In my opinion, this PR is missing three tests before getting accepted:
1. Make sure that a GET /subsidies returns 200 in all cases -- no records or many records.
2. I'm pretty sure with the code written this way, a PUT request with a negative value (-13) will crash the NodeJS server.  You should write a test to prove me wrong.
3. PUT with an ID in the object submitted should not rewrite the ID in the database.

File: server/controllers/finance/subsidy.js (hunk 2 above)
Comment (DedrickEnc) [inline server/controllers/finance/subsidy.js:107]: The test is clear, it is less than zero, so that value is not correct.

Comment (jniles) [inline server/controllers/finance/subsidy.js:107]: The test is very clear.  The problem is that an API consumer shouldn't have to have the application source code to know why an error occurred.  A helpful error message would save a lot of time and pain for that developer.

File: server/controllers/finance/subsidy.js (hunk 3 above)
Comment (DedrickEnc) [inline server/controllers/finance/subsidy.js:106]: As many PR is using this ERR_PARAMETERS_REQUIRED property, it can be corrected as separated pull request.

File: server/test/api/subsidy.js (hunk 4 above)
Comment (DedrickEnc) [inline server/test/api/subsidy.js]: you are right

File: server/controllers/finance/subsidy.js (hunk 3 above)
Comment (jniles) [inline server/controllers/finance/subsidy.js:106]: Sounds good.

File: server/test/api/subsidy.js (hunk 6 above)
Comment (DedrickEnc) [inline server/test/api/subsidy.js:43]: Be more clear, please!

Comment (jniles) [inline server/test/api/subsidy.js:43]:
``` js
// just after the function definition for `describe()`
var keys = ['id', 'account_id', 'label', 'description', 'value', 'created_at', 'updated_at'];

// later in a  test
expect(res.body).to.have.all.keys(keys);
```

File: server/test/api/subsidy.js (hunk 8 above)
Comment (DedrickEnc) [inline server/test/api/subsidy.js:71]: I understand, but no matter, it is more informative, here the `/subsidies/:id` is tested directly.
Each route must be tested, directly and separately, `subsidies/:id` is one of them.

File: server/test/api/subsidy.js (hunk 9 above)
Comment (DedrickEnc) [inline server/test/api/subsidy.js:93]: Why?

File: server/test/api/subsidy.js (hunk 11 above)
Comment (DedrickEnc) [inline server/test/api/subsidy.js:29]: For common issue like this, a separate pull request will be good, because almost all our test is dealing like this, so a separate pull request will walk through all those file and correct them if necessary.

File: server/test/api/subsidy.js (hunk 9 above)
Comment (jniles) [inline server/test/api/subsidy.js:93]: Why not?

File: server/test/api/subsidy.js (hunk 8 above)
Comment (jniles) [inline server/test/api/subsidy.js:71]: Duplicate tests aren't a terrible problem.  One caution though: if the POST method breaks, you may spend lots of time trying to debug the GET request rather than the POST request, since it will look like the GET request is failing as well.

**Comment by DedrickEnc** [comment]
@jniles 
For the : 
1. I think a separate pull request will be good, because it is a common situation.
2. Since I made a test with a negative value, I don't see why you think that -13 will break the server?
3. I am deleting the id before performing an update request, just see the code.

File: server/test/api/subsidy.js (hunk 11 above)
Comment (jniles) [inline server/test/api/subsidy.js:29]: I'll fix the other tests in the `test-handlers` PR if you'll do this one, that would be very helpful.

**Comment by DedrickEnc** [comment]
Sorry, you are right for the PUT with negative value

**Comment by DedrickEnc** [comment]
Before updating a value, I am calling the function `checkData`

**Comment by jniles** [comment]
Can you prove to me that the server will not break?  I'd like to see a passing test for the UPDATE with a negative value.  The others are just comments to future-proof the code (so that changes to the way the API works are immediately noticed).

File: server/test/api/subsidy.js (hunk 10 above)
Comment (DedrickEnc) [inline server/test/api/subsidy.js]: That will lead me to duplicate the test, which is not profitable

**Comment by DedrickEnc** [comment]
@jniles you can review it again

**Comment by jniles** [comment]
Normally,  `checkData()` should crash the process.  It uses a `throw` statement without being in the context of a `try`/`catch` block, or in a promise.  However, it doesn't because of weird behavior in NodeJS.

It turns out that there is a difference in errors are created in Chrome and NodeJS.  It is unclear to me why this is the case, since it contradicts the [MDN documentation for throw](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/throw).

``` js
// in the chrome browser, all of these generate an exception
throw new Error('This is an error from a String');
throw Error('This also works, without the new statement');
throw 'In fact, you don\'t need the Error function, you can just write a string...  No call stack though.';

// in NodeJS the two following generate an exception and crash the process
throw new Error('This is an error from a String');
throw Error('This also works, without the new statement');
// ... but this doesn't!
throw 'Why doesn\'t NodeJS abide by JavaScript standards?';
```

I will accept this PR since it doesn't crash the process like I thought (I pulled your branch locally and wrote two more integration tests to try it).

However, in the future, because of NodeJS's non-standard behavior, I will change `req.codes` to export Error objects instead of plain objects, which will break this PR.  I'll add in the new tests in that PR and make sure that the process doesn't crash.

@dedrickenc, could you do a `rebase -i` to master and squash the two extra commits ("amending")?  It will make our commit history much cleaner.

**Comment by jniles** [comment]
LGTM.  Thanks!
